search_backend/api/lib/indexing_pipeline.py

```diff
@@ -0,0 +1,82 @@
+from typing import Iterable, Any
+
+from haystack import Pipeline, Document
+from haystack.components.preprocessors import DocumentSplitter
+from haystack.components.writers import DocumentWriter
+from haystack.document_stores.types import DuplicatePolicy
+from haystack_integrations.components.embedders.fastembed import FastembedDocumentEmbedder
+from haystack_integrations.document_stores.opensearch import OpenSearchDocumentStore
+
+
+class IndexingPipeline:
+    """
+    Index documents into opensearch document store.
+    """
+
+    def __init__(
+            self,
+            document_store: OpenSearchDocumentStore,
+            dense_embedding_model: str,
+            semantic: bool = False,
+            indexing: Pipeline = Pipeline(),
+    ):
+        """
+        :param document_store: DocumentStore object that has been set up elsewhere
+        :param dense_embedding_model: name of the dense embedding model
+        :param semantic: set this to True to enable semantic/hybrid search. Otherwise uses a BM25 search. If True,
+        the chunks of text are embedded and written to a vector store.
+        :param indexing: pipeline to do the indexing, which will be configured in this constructor
+        """
+        self.document_store = document_store
+
+        document_splitter = DocumentSplitter(split_length=128, split_overlap=32)
+
+        indexing.add_component("document_splitter", document_splitter)
+
+        indexing.add_component(
+            "document_writer",
+            DocumentWriter(document_store=self.document_store, policy=DuplicatePolicy.OVERWRITE),
+        )
+
+        if semantic:
+            indexing.add_component(
+                "dense_doc_embedder",
+                FastembedDocumentEmbedder(model=dense_embedding_model),
+            )
+
+            indexing.connect("document_splitter", "dense_doc_embedder")
+            indexing.connect("dense_doc_embedder", "document_writer")
+        else:
+            indexing.connect("document_splitter", "document_writer")
+
+        self.indexing = indexing
+
+    def run(self, docs: Iterable[Document]):
+        """
+        Split the data into chunks and write it to the document store.
+
+        :param docs: Haystack Document objects to be indexed.
+        """
+        return self.indexing.run({"document_splitter": {"documents": docs}})
+
+    def delete_docs(self, document_ids: list[Any], id_metafield: str):
```

Comment:
Do we want to do anything defensive for document_ids, like expect a set? I presume we don't want duplicated document IDs.

Reply:
It doesn't make any difference to haystack, but it wouldn't hurt.

Reply:
I just checked, and the IDs are turned into a set later. Bear in mind that these are filter criteria, not actually IDs, so might potentially contain duplicates (depending on where the filters come from). After filtering, the retrieved document IDs are turned into a set, so we will only attempt to remove each document once.